packages/api/src/clickhouse/index.ts

```diff
@@ -904,6 +912,588 @@ export const getMetricsChart = async ({
+        : aggFn === AggFn.SumRate
+        ? `SUM(rate) as data`
+        : aggFn === AggFn.AvgRate
+        ? `AVG(rate) as data`
+        : aggFn === AggFn.MaxRate
+        ? `MAX(rate) as data`
+        : aggFn === AggFn.MinRate
+        ? `MIN(rate) as data`
+        : `quantile(${
+            aggFn === AggFn.P50 || aggFn === AggFn.P50Rate
+              ? '0.5'
+              : aggFn === AggFn.P90 || aggFn === AggFn.P90Rate
+              ? '0.90'
+              : aggFn === AggFn.P95 || aggFn === AggFn.P95Rate
+              ? '0.95'
+              : '0.99'
+          })(${isRate ? 'rate' : 'value'}) as data`,
+    );
+  } else {
+    logger.error(`Unsupported data type: ${dataType}`);
+  }
+
+  // used to sum/avg/percentile Sum metrics
+  // max/min don't require pre-bucketing the Sum timeseries
+  const sumMetricSource = SqlString.format(
+    `
+      SELECT
+        toStartOfInterval(timestamp, INTERVAL ?) as timestamp,
+        min(value) as value,
+        _string_attributes,
+        name
+      FROM ??
+      WHERE name = ?
+      AND data_type = ?
+      AND (?)
+      GROUP BY
+        name,
+        _string_attributes,
+        timestamp
+      ORDER BY
+        _string_attributes,
+        timestamp ASC
+    `.trim(),
+    [granularity, tableName, name, dataType, SqlString.raw(whereClause)],
+  );
+
+  const rateMetricSource = SqlString.format(
+    `
+      SELECT
+        if(
+          runningDifference(value) < 0
+          OR neighbor(_string_attributes, -1, _string_attributes) != _string_attributes,
+          nan,
+          runningDifference(value)
+        ) AS rate,
+        timestamp,
+        _string_attributes,
+        name
+      FROM (?)
+      WHERE isNaN(rate) = 0
+      ${shouldModifyStartTime ? 'AND timestamp >= fromUnixTimestamp(?)' : ''}
+    `.trim(),
+    [
+      SqlString.raw(sumMetricSource),
+      ...(shouldModifyStartTime ? [startTime / 1000] : []),
+    ],
+  );
+
+  const gaugeMetricSource = SqlString.format(
+    `
+      SELECT
+        toStartOfInterval(timestamp, INTERVAL ?) as timestamp,
+        name,
+        last_value(value) as value,
+        _string_attributes
+      FROM ??
+      WHERE name = ?
+      AND data_type = ?
+      AND (?)
+      GROUP BY name, _string_attributes, timestamp
+      ORDER BY timestamp ASC
+    `.trim(),
+    [granularity, tableName, name, dataType, SqlString.raw(whereClause)],
+  );
+
+  const query = SqlString.format(
+    `
+      WITH metrics AS (?)
+      SELECT ?
+      FROM metrics
+      GROUP BY group, ts_bucket
+      ORDER BY ts_bucket ASC
+      ${
+        granularity != null
+          ? `WITH FILL
+        FROM toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        TO toUnixTimestamp(toStartOfInterval(toDateTime(?), INTERVAL ?))
+        STEP ?`
+          : ''
+      }
+    `,
+    [
+      SqlString.raw(
+        isRate
+          ? rateMetricSource
+          : // Max/Min aggs are the same for both Sum and Gauge metrics
```

Comment:
Not sure I fully understand this. So if Mas/Min aggs are the same for both, why do we need extra aggFn checking here ?

Reply:
This is moved from the existing [`getMetricsChart` function](https://github.com/hyperdxio/hyperdx/blob/72164a64a8bfd9591e8629a656f57cac39033195/packages/api/src/clickhouse/index.ts#L866);

In a future PR I actually intend to get rid of this condition, since the `sumMetricSource` and `gaugeMetricSource` are nearly identical after the agg fn bug fix I talked about on Slack.

This logic was here because the min/max aggFn didn't require the more expensive `sumMetricSource` query, even for sum type metrics. However, the `gaugeMetricSource` and `sumMetricSource` will in the future be the same query in general, so this shouldn't really matter.